contracts/btc-staking/src/state/delegations.rs

```diff
@@ -0,0 +1,102 @@
+use crate::state::points_alignment::PointsAlignment;
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{CanonicalAddr, Order, StdResult, Storage, Uint128};
+use cw_storage_plus::{Index, IndexList, IndexedMap, KeyDeserialize, MultiIndex};
+
+/// Single delegation related information - entry per `(staking hash, finality provider public key)`
+/// pair, including distribution alignment
+#[cw_serde]
+pub struct Delegation {
+    /// The delegator's canonical address
+    pub staker_addr: CanonicalAddr,
+    /// How many satoshis the user stakes in this delegation
+    pub stake: u64,
+    /// How many points should be added / subtracted from points calculated per delegation due to
+    /// delegation changes.
+    pub points_alignment: PointsAlignment,
+    /// Rewards already withdrawn by this user
+    pub withdrawn_funds: Uint128,
+}
+
+#[allow(clippy::type_complexity)]
+pub struct DelegationIndexes<'a> {
+    // Delegations by finality provider's public key and staking hash.
+    // Last type param defines the pk deserialization type
+    pub rev: MultiIndex<'a, (String, Vec<u8>), Delegation, (Vec<u8>, String)>,
+    // Delegations by staker's (raw, canonical) address and finality provider's public key.
+    // Last type param defines the pk deserialization type
+    pub staker: MultiIndex<'a, (Vec<u8>, String), Delegation, (Vec<u8>, String)>,
```

Comment:
Not sure if I understand these two multi index:

- Why would the key contain both raw and canonical addresses? Probably raw address is good enough as the reward distribution does not care about the account prefix at all?
- What is the definition of `rev`?
- Why should we use multi index and differentiate key / primary key?

Reply:
The key does not contain both, just raw.

`rev` is a reverse index of delegations per FP. It can be used to replace the `DELEGATIONS_FPS` store entirely. Same with the direct key, which can be used to replace the current `FP_DELEGATIONS`.

There's already a TODO in the code for that. Will address in a follow-up.